Mathlib/MeasureTheory/Measure/Tilted.lean

```diff
@@ -0,0 +1,367 @@
+open Real
+
+open scoped ENNReal NNReal
+
+namespace MeasureTheory
+
+variable {α : Type*} {mα : MeasurableSpace α} {μ : Measure α} {f : α → ℝ}
+
+/-- Exponentially tilted measure. When `x ↦ exp (f x)` is integrable, `μ.tilted f` is the
+probability measure with density with respect to `μ` proportional to `exp (f x)`. Otherwise it is 0.
+-/
+noncomputable
+def Measure.tilted (μ : Measure α) (f : α → ℝ) : Measure α :=
+  μ.withDensity (fun x ↦ ENNReal.ofReal (exp (f x) / ∫ x, exp (f x) ∂μ))
+
+attribute [pp_dot] Measure.tilted
+
+@[simp]
+lemma tilted_of_not_integrable (hf : ¬ Integrable (fun x ↦ exp (f x)) μ) : μ.tilted f = 0 := by
+  rw [Measure.tilted, integral_undef hf]
+  simp
+
+@[simp]
+lemma tilted_of_not_aemeasurable (hf : ¬ AEMeasurable f μ) : μ.tilted f = 0 := by
+  refine tilted_of_not_integrable ?_
+  suffices ¬ AEMeasurable (fun x ↦ exp (f x)) μ by exact fun h ↦ this h.1.aemeasurable
+  exact fun h ↦ hf (aemeasurable_of_aemeasurable_exp h)
+
+@[simp]
+lemma tilted_zero_measure (f : α → ℝ) : (0 : Measure α).tilted f = 0 := by simp [Measure.tilted]
+
+@[simp]
+lemma tilted_const' (μ : Measure α) (c : ℝ) :
+    μ.tilted (fun _ ↦ c) = (μ Set.univ)⁻¹ • μ := by
+  cases eq_zero_or_neZero μ with
+  | inl h => rw [h]; simp
+  | inr h0 =>
+    simp only [Measure.tilted, withDensity_const, integral_const, smul_eq_mul]
+    by_cases h_univ : μ Set.univ = ∞
+    · simp only [h_univ, ENNReal.top_toReal, zero_mul, log_zero, div_zero, ENNReal.ofReal_zero,
+        zero_smul, ENNReal.inv_top]
+    congr
+    rw [div_eq_mul_inv, mul_inv, mul_comm, mul_assoc, inv_mul_cancel (exp_pos _).ne', mul_one,
+      ← ENNReal.toReal_inv, ENNReal.ofReal_toReal]
+    simp [h0.out]
+
+lemma tilted_const (μ : Measure α) [IsProbabilityMeasure μ] (c : ℝ) :
+    μ.tilted (fun _ ↦ c) = μ := by simp
+
+@[simp]
+lemma tilted_zero' (μ : Measure α) : μ.tilted 0 = (μ Set.univ)⁻¹ • μ := by
+  change μ.tilted (fun _ ↦ 0) = (μ Set.univ)⁻¹ • μ
+  simp
+
+lemma tilted_zero (μ : Measure α) [IsProbabilityMeasure μ] : μ.tilted 0 = μ := by simp
+
+lemma tilted_congr {g : α → ℝ} (hfg : f =ᵐ[μ] g) :
+    μ.tilted f = μ.tilted g := by
+  have h_int_eq : ∫ x, exp (f x) ∂μ = ∫ x, exp (g x) ∂μ := by
+    refine integral_congr_ae ?_
+    filter_upwards [hfg] with x hx
+    rw [hx]
+  refine withDensity_congr_ae ?_
+  filter_upwards [hfg] with x hx
+  rw [h_int_eq, hx]
+
+lemma tilted_eq_withDensity_nnreal (μ : Measure α) (f : α → ℝ) :
+    μ.tilted f = μ.withDensity (fun x ↦ ((↑) : ℝ≥0 → ℝ≥0∞)
+      (⟨exp (f x) / ∫ x, exp (f x) ∂μ,
+      div_nonneg (exp_pos _).le (integral_nonneg (fun _ ↦ (exp_pos _).le))⟩ : ℝ≥0)) := by
+  rw [Measure.tilted]
+  congr with x
+  rw [ENNReal.ofReal_eq_coe_nnreal]
+
+lemma tilted_apply (μ : Measure α) (f : α → ℝ) {s : Set α} (hs : MeasurableSet s) :
+    μ.tilted f s = ∫⁻ a in s, ENNReal.ofReal (exp (f a) / ∫ x, exp (f x) ∂μ) ∂μ := by
+  rw [Measure.tilted, withDensity_apply _ hs]
+
+lemma tilted_apply' (μ : Measure α) [SFinite μ] (f : α → ℝ) (s : Set α) :
```

Comment:
don't you want to exchange the primed and non-primed version? I expect most of the time this will be used with probability measures. Same thing below.